### PR TITLE
changes the means to compute the version for packages

### PR DESCRIPTION
**Pull request by cmickeyb**
Change the way we compute the version so that github workflows can use this

Version information is kept in the file VERSION. The file contains lines for each version from the most recent to the most distant. Each tab separated line contains a version (of the form v?[0-9]+.[0-9]+.[0-9]+) and the hash of a git commit that represents that version. Each line may also have a short description separated by a tab.

When a new version is committed, only the VERSION file should be modified and the commit should be tagged (annotated tag) with the version number.

### CODE REVIEW

**Comment by g2flyer** [comment]
> When a new version is committed, only the VERSION file should be modified and the commit should be tagged (annotated tag) with the version number.

Hmm, if we tag that version-update commit, why do we still need a VERSION file and that commit and do not just use the tag? Seems to me we really should use only one or the other (VERSION file or tag) to reduce confusion (in particular given that commit in VERSION file is by definition (at least) one commit version before the tag?)   Also, if we maintain the VERSION file, maybe it would be good to have a script which updates it based on current commit and a version provided to script and right away commits?  Besides being convenient it would ensure consistency?

**Comment by cmickeyb** [comment]
> > When a new version is committed, only the VERSION file should be modified and the commit should be tagged (annotated tag) with the version number.
> 
> Hmm, if we tag that version-update commit, why do we still need a VERSION file and that commit and do not just use the tag? Seems to me we really should use only one or the other (VERSION file or tag) to reduce confusion (in particular given that commit in VERSION file is by definition (at least) one commit version before the tag?) Also, if we maintain the VERSION file, maybe it would be good to have a script which updates it based on current commit and a version provided to script and right away commits? Besides being convenient it would ensure consistency?

the version file should be authoritative. the version tag should just be for retrieving the commit where the version file changed. this can be automated (in theory) with a github workflow. in practice, this appears harder than it looks.

**Comment by g2flyer** [comment]
> > > When a new version is committed, only the VERSION file should be modified and the commit should be tagged (annotated tag) with the version number.
> > 
> > 
> > Hmm, if we tag that version-update commit, why do we still need a VERSION file and that commit and do not just use the tag? Seems to me we really should use only one or the other (VERSION file or tag) to reduce confusion (in particular given that commit in VERSION file is by definition (at least) one commit version before the tag?) Also, if we maintain the VERSION file, maybe it would be good to have a script which updates it based on current commit and a version provided to script and right away commits? Besides being convenient it would ensure consistency?
> 
> the version file should be authoritative. the version tag should just be for retrieving the commit where the version file changed. this can be automated (in theory) with a github workflow. in practice, this appears harder than it looks.

Creating a script (parameterized by new version) in repo which updates VERSION, commits & tags could be an easy poor-man's version of the github workflow?

**Comment by cmickeyb** [comment]
> Creating a script (parameterized by new version) in repo which updates VERSION, commits & tags could be an easy poor-man's version of the github workflow? No matter what, a version update has to be executed manually so calling that script in your local repo is arguably even simpler than going to the repo web UI and trigger some workflow?

Its reasonable to have a script to create the version. But there is a catch-22 here. You can't tag a version that you don't have. The workflow is probably the right way to do it in the long run. In the short term, we have to have a legitimate version. And everything we currently build in github is 0.0.0 (which means there is no way to use GHCR to store and retrieve the images).

**Comment by g2flyer** [comment]
> Its reasonable to have a script to create the version. But there is a catch-22 here. You can't tag a version that you don't have. The workflow is probably the right way to do it in the long run. In the short term, we have to have a legitimate version. And everything we currently build in github is 0.0.0 (which means there is no way to use GHCR to store and retrieve the images).

Not sure i understand: what i was thinking is something along the lines of your set_version which updates the version file _but_ then also does right away the commit (because as you said in desc and i agree that version update should be in standalone commit); in that case you could tag afterwards (and as long as you do a merge commit in PR or a rebase from top-of-master/main then the tag should be the right now?

File: bin/get_version [hunk 1]
```diff
@@ -14,35 +14,42 @@
...
-    print('0.0.0')
+
+pdo_source_root=pathlib.Path(__file__).parent.parent
+version_file = pdo_source_root / 'VERSION'
+
+parser = argparse.ArgumentParser()
+
+parser.add_argument(
+    '--version-file', '-f',
+    help='File where version information is stored',
```
Comment (g2flyer) [inline bin/get_version:29]:
```suggestion
    help=f'File where version information is stored (default: {version_file})',
```

File: bin/set_version [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+# Add a new version to the version file. The version file is a tab separated list of version numbers
+# and git commit hashes in reverse order (newest is at the top of the file). The version may contain
+# a short description as well.
+
+import argparse
+import datetime
+import os
+import pathlib
+import subprocess
```
Comment (g2flyer) [inline bin/set_version:25]:
```suggestion
import subprocess
import sys
```

File: bin/set_version [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        print(f'Invalid version string; {version}')
+        sys.exit(1)
+
+    minor = int(minor) + 1
+    patch = 0
+
+# Compute the commit to associate with the new version
+if options.commit :
+    command = ['git', 'rev-parse', options.commit]
+    output = subprocess.run(command, cwd=pdo_source_root, capture_output=True, text=True)
```
Comment (g2flyer) [inline bin/set_version:87]: we should check exit code and abort on error. Right now if you provide a non-existing commit (or tag) with `-c` the version file gets updated with `<non-existing-commit>   <today's-date>`

**Review by g2flyer — COMMENTED** [review]
A few minor suggestion and one bug fix. 

PS: besides playing around with the `*_version` scripts (where i found the one bug) i also tested docker and all that worked as expected.

File: bin/set_version [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    type=str)
+
+options = parser.parse_args()
+
+# Get the VERSION file and current source path
+if options.version_file :
+    version_file = pathlib.Path(options.version_file)
+    pdo_source_root = version_file.parent
+
+# Compute the new version, by default the version will increment the minor version of the most
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply bin/set_version:59]: actually i'm saying that we generally only commit a new (tagged) version on improvement. this is patently wrong, but if you want more control then just set the version explicitly.

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -8,7 +8,7 @@ jobs:
     pdo_ci:
         if: "!contains(github.event.commits[0].message, '[skip ci]')"
         name: PDO CI Job
-        runs-on: ubuntu-20.04
+        runs-on: ubuntu-22.04
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply .github/workflows/ci.yaml:11]: remember that this is the version only used for building the docker images. nothing else runs here. so its not actually related to our documented default version. i could go for a jump to 24.04 if that actually makes a difference. only wanted this for improved docker commands.

File: bin/set_version (hunk 3 above)
Comment (cmickeyb) [reply bin/set_version:87]: will add this

Comment (g2flyer) [reply bin/set_version:88]: Seems issue is that there is a missing `, check=True ` among the subprocess.run arguments to get an exception. Once you add that it does the right thing ...

**Review by g2flyer — APPROVED** [review]

**Pull request merged**